### ISSUE TITLE
Standardized Variable Names and IEEE NaN as a fill value

### ISSUE THREAD

**Issue by DocOtak** [cchdo/hydro#3]
These discussions started over email due to bad bandwidth, coping here now that my ship is back on fast internet, usernames modified to be the github usernmae:

Hi @efiring and @ocefpaf,

CCHDO is in the process of designing a CF-Compliant netCDF file, drafts attached.

One of the decisions made was to have no standardized variable names and rely completely on variable attributes to sort out what is what in the file. With certain variables which describe coordinates as exceptions. Basically, it is a combination of Argo prof files, and the output of ODV when it exports a netcdf file, with a hopefully minimal set of modifications needed to add CCHDO specific metadata.

Some background on the decision to not have standardized variable names:
* Current WHP exchange file names are mostly tied to the physical property being reported, with no unit information (it’s on a separate line).
* Most implementations expect the “Parameter” line to contain unique comma separated values.
* What to name new parameters in exchange files has been the source of significant time wasted and argument, often some of our… experienced users will ask for new names that are at most 6 chars.
* The CF document explicitly states it does not standardize variable names only attributes and their values (and a little bit of geometry).

Personal anecdote: when things are explicitly “not standardized” but there are some consistencies in one particular implementation, users will become reliant on those consistencies, and I’m keen to avoid this situation.

So, given the above, users who primarily use MatLab have expressed concerns over the nonstandization of the variable names. I think these concerns would be largely mitigated if there existed some form of python’s get_variables_by_attributes for matlab. In my searching, I haven’t been able to find much. I’d imagine a library for working with CF data might work as well.

Thanks,
Andrew Barna

**Comment by DocOtak** [comment]
From @ocefpaf
---
Sorry Andrew but I don't use matlab too much lately and I'm not even sure what is the best way to read netcdf nowadays.
What I do know is that it used to be a wrapper to the java library and the `get_variables_by_attributes` method was based on a Java library method,
so maybe that exists already.

Good luck,

-F

**Comment by DocOtak** [comment]
From @efiring
----
Andrew,

Sorry to be slow in responding, but I don't think there is anything I can contribute.  If it hasn't already been done, it looks like someone will need to write a matlab toolkit that provides a nicer netcdf interface than is found in either matlab's high-level or low-level sets of functions.  Or, matlab people will just have to use ncdump -h to figure out what "varxx" they need.

Overall, this file format looks pretty clunky and painful to use; is it really necessary to include so many attributes for each variable?

Maybe it would be possible to add a "table of contents", a single attribute or variable that would map standard names to varxx.

Eric

**Comment by DocOtak** [comment]
From @DocOtak 
---
Eric,

The mess of attributes you see is the result of pushing down the ACDD attributes from the global to the variable level. This is largely in response to how the programs/collaborations (GO-SHIP is the big one) we manage data for are set up. Where some collection of variables has one PI and some other collection of variables has a different one, e.g. the CTDO part of the CTD one PI (Jim Swift), but the transmissometer has a different PI (Wilf), and even the FLBB has a different PI. That’s just the CTD, many PI’s are involved with the bottle data.

This file was an attempt to solve the questions our programs PIs and our program managers ask us:

* Where was variable X measured? (Often different locations from variable Y)
* Who was the PI for variable X?
* When did they turn in their data if they have?
* When is the data for variable X expected? (GO-SHIP sets timelines for this)
* I have some specific question for the techs who did the measurement on board (SOCCOM has asked this, but the specific tech was me)…
* Are these data “final”?

Then there is all the CF bookkeeping:
* Standard name
* Units
* FLag values and meanings
* Linking any ancillary variables (e.g. flag vars for this parameter)
* Linking up auxiliary coordinates (e.g. woce stnnbr,castno,sampno or the geotraces numbers)

We are also including the woce names for variables when they are present, e.g. CTDTMP and ITS-90 for the thing with the standard name of “sea_water_temperature”.

To remain compliant with ACDD, all of their attributes are combined into long lists in the global attributes.

Other things going on:
* CCHDO is attempting to remove itself as a naming authority, the current plan is to eventually “freeze” the list of whp exchange format parameter names, and any new measurements would only be available in a netcdf file with associated CF standard name, suggesting one using their rules if it does not already exist.
* An extension to the above, we also don’t want to be coming up with standard or short names for variable names, and figured the way ODV outputs netcdf was “good enough for oceanographers”
* We want the geometry of the data to match argo as close as we can.

The above is just the reasoning behind what you see, I don’t intend it just yet to be the justification.

Some specific questions:
Is the geometry of the data ok?
Is it just the massive amount of attributes which make this “pretty clunky” or is it something else?

Thanks for your feedback.

-Barna

**Comment by DocOtak** [comment]
From @efiring
---
>Some specific questions:
>Is the geometry of the data ok?

Would it be desirable to make a coordinate array for the N_WAVELENGTH0 dimension, at least?  It would be a matter of using the same name for the dimension and the array, which is presently wavelength0. (This is motivated by loading the file with xarray; but I'm not sure how much advantage there is in having coordinate arrays for each dimension.)  I don't think I've seen this "N_SOMETHING" convention before for dimensions; I would have expected dimensions profile, level, and wavelength0, with corresponding arrays of the same name, even if some of them are just sequential integers.

I'm posing the above as a question because I really don't know the answer, not because I think things should be done differently.

> Is it just the massive amount of attributes which make this “pretty clunky” or is it something else?

Both the barrage of attributes and the use of "var01" etc. make the file awkward to approach.

You have some variables like "var6" that don't have a standard name at all, just a whp_name.  Why not give them real names, like "WHP_EXPOCODE"?  (Side note: xarray apparently ignores them; maybe it doesn't handle string arrays?)

I was also wondering about the use of NaN as a _FillValue.  This seems to me a bad practice because it means that code testing for the locations of the _FillValue requires 2 code paths, one to handle the NaN case, and another for the normal case in which a normal value of the correct type is designated.

What do you think of my suggestion to add a "table of variables" in some form?

Eric

**Comment by DocOtak** [comment]
From @DocOtak 
---
Thanks Eric,

> >Some specific questions:
> > Is the geometry of the data ok?
> 
> Would it be desirable to make a coordinate array for the N_WAVELENGTH0 dimension, at least?  It would be a matter of using the same name for the dimension and the array, which is presently wavelength0. (This is motivated by loading the file with xarray; but I'm not sure how much advantage there is in having coordinate arrays for each dimension.)  I don't think I've seen this "N_SOMETHING" convention before for dimensions; I would have expected dimensions profile, level, and wavelength0, with corresponding arrays of the same name, even if some of them are just sequential integers.

Yes, probably could just make the wavelength one a Coordinate Variable (in the NUG definition of the term), while n=3 is a less than optimal example of why this might be desirable, it is being set up for what are essentially profiling spectrometers where the number of discrete wavelengths is in the 100s to 1000s. Different instruments will need different numbers of wavelengths and values, e.g. we might have a case where there are wavlength0, wavelength1, ... wavelengthN for some optically heavy data.

See the argo “_prof.nc” files for an example of “N_SOMETHING”, we took the N_PROF and N_LEVEL directly from the argo profile 3.2 netcdf file, as defined in table 2.2.2 http://dx.doi.org/10.13155/29825


> I'm posing the above as a question because I really don't know the answer, not because I think things should be done differently.
> 
> > Is it just the massive amount of attributes which make this “pretty clunky” or is it something else?
> 
> Both the barrage of attributes and the use of "var01" etc. make the file awkward to approach.

Would written for humans documentation help? This would probably include code example of how to get “your favorite view” out of a verbose file… In my personal experience, attempting to “feel out” a file, even one that is supposed to be “self describing” is going to result in errors. Examples I can think of: someone on board is trying to read csv files as fixed width and getting extremely frustrated with “commas jumping around”; you are supposed to use “_adjusted" values in argo files over the unadjusted ones, but the files don’t tell you that


> You have some variables like "var6" that don't have a standard name at all, just a whp_name.  Why not give them real names, like "WHP_EXPOCODE"?  (Side note: xarray apparently ignores them; maybe it doesn't handle string arrays?)

I think this would only really work for the WHP sample identifying information, these would be the compound keys consisting of EXPOCODE, STNNBR, CASTNO, and SAMPNO. Note that despite the “NO”s and “NBR”s in the names, only CASTNO is defined to be a numeric value by the woce manual, everything else can have “alphanumeric” values.

A good compromise on this would be “anything which can be considered a coordinate” can be named at the variable level: lat, lon, time, (expocode, stn, cast, sample), (geotraces number, geotraces sample). I use “coordinate” here to mean some value which can identify the data it labels in spacetime or within a program, not in the NUG Coordinate Variable sense.

CTD data in WOCE do not have a non-measured value in the Z axis, combining data from sensors not in the same data stream has thus far been done by ensuring that we are all using the same pressure bins.

I would want to avoid including measured parameters as variable names since we have the issue of multiple parallel measurements and coming up with new names has proven to be an expensive task. Example: on the recent geotraces cruise on the Revelle, there were up to 3 PIs measuring the same physical quantities, off of 5 different sampling systems. In the case of Iron, this resulted in 22 different parameter names capturing the PI, sampling system, physical property, etc.. This resulted in names such as the following: Fe_D_CONC_BOTTLE::HATTA, meaning “Iron, dissolved, off a niskin/go-flo bottle, Hatta as PI”, the specific sampling system is in the BTLNBR column (also not a number). It just seemed the only way to reasonably capture all this information is via attributes, cramming it into a single “name” results in something worse than too many attributes.

Total parameter name count on that geotraces cruise was over 500...


> I was also wondering about the use of NaN as a _FillValue.  This seems to me a bad practice because it means that code testing for the locations of the _FillValue requires 2 code paths, one to handle the NaN case, and another for the normal case in which a normal value of the correct type is designated.

Use of NaNs for floating point data as recommended by https://www.unidata.ucar.edu/software/netcdf/workshops/2011/bestpractices/MissingData.html

xarray appears to coerce non contiguous numeric data into a float type and encode the actual file dtype information for round trip purposes. I don’t know if this is xarray behavior or if the underlying netcdf library forces something like this to happen.

When I was going though the CF mailing list archives for recommendations on fill values, it seemed that the usual practice is to take the approach that xarray does, turn all numeric data into something which follows IEEE 754 rules, then do all the testing/math using nan aware logic. Missing data is a little complicated in netcdf files, for example, I could use the NUG valid_range attribute and not include a _FillValue, then ensure that all “non valid” data remains outside the valid range.

Basically, every approach to missing data values seems to a compromise that no one is happy with.

>  What do you think of my suggestion to add a "table of variables" in some form?

I’m having trouble picturing what this would actually look like in the constraints of a netcdf file...
An array of strings containing values like: “CTDPRS:var0” “CTDSAL:var2”
Something else maybe?

Thanks Again for your feedback,
Barna

**Comment by DocOtak** [comment]
From @efiring
---

Single string, or array of strings; there are many possible variations (quotes or no quotes?  commas? newlines? spaces?), with different tradeoffs in terms of printing nicely in ncdump (which would probably require a separate attribute for each variable--not so nice--given ncdump's present behavior with netcdf4 format.), or in other interfaces, and ease of using it programmatically to make a dictionary, for example.

The script attached gives two examples.  I put the var first because it will make for more readable tables, given how horrendously long some of the "standard names" are.

I will use a separate message to comment on some of your other points.

Eric

```python
import netCDF4 as nc

fname = 'test_example_cchdo.nc'

d = nc.Dataset(fname, 'r+')

varlist = []

for name, var in d.variables.items():
    if not name.startswith('var'):
        continue
    if 'standard_name' not in var.ncattrs():
        continue
    varlist.append((name, var.getncattr('standard_name')))

# There are many possible variations on the following formatting:
strlist = [f'   "{vn}" : "{sn}",\n' for vn, sn in varlist]

toc = ''.join(strlist)

# List of strings?
d.setncattr('variable_names1', strlist)
# Or a single string?
d.setncattr('variable_names2', toc)

d.close()
```

**Comment by DocOtak** [comment]
From @efiring
---

> Would written for humans documentation help? This would probably include code example of how to get “your favorite view” out of a verbose file… In my personal experience, attempting to “feel out” a file, even one that is supposed to be “self describing” is going to result in errors. Examples I can think of: someone on board is trying to read csv files as fixed width and getting extremely frustrated with “commas jumping around”; you are supposed to use “_adjusted" values in argo files over the unadjusted ones, but the files don’t tell you that

Yes, documentation would help.


> > You have some variables like "var6" that don't have a standard name at all, just a whp_name.  Why not give them real names, like "WHP_EXPOCODE"?  (Side note: xarray apparently ignores them; maybe it doesn't handle string arrays?)
>
> I think this would only really work for the WHP sample identifying information, these would be the compound keys consisting of EXPOCODE, STNNBR, CASTNO, and SAMPNO. Note that despite the “NO”s and “NBR”s in the names, only CASTNO is defined to be a numeric value by the woce manual, everything else can have “alphanumeric” values.

Yes, an unfortunate choice that we are stuck with.  But it doesn't affect my basic point which is to simplify the file by minimizing the number of "var*" names.  (I also dislike manufacturers "serial numbers" that are in fact alphanumeric.)

> A good compromise on this would be “anything which can be considered a coordinate” can be named at the variable level: lat, lon, time, (expocode, stn, cast, sample), (geotraces number, geotraces sample). I use “coordinate” here to mean some value which can identify the data it labels in spacetime or within a program, not in the NUG Coordinate Variable sense. >
> CTD data in WOCE do not have a non-measured value in the Z axis, combining data from sensors not in the same data stream has thus far been done by ensuring that we are all using the same pressure bins.
> I would want to avoid including measured parameters as variable names since we have the issue of multiple parallel measurements and coming up with new names has proven to be an expensive task. Example: on the recent geotraces cruise on the Revelle, there were up to 3 PIs measuring the same physical quantities, off of 5 different sampling systems. In the case of Iron, this resulted in 22 different parameter names capturing the PI, sampling system, physical property, etc.. This resulted in names such as the following: Fe_D_CONC_BOTTLE::HATTA, meaning “Iron, dissolved, off a niskin/go-flo bottle, Hatta as PI”, the specific sampling system is in the BTLNBR column (also not a number). It just seemed the only way to reasonably capture all this information is via attributes, cramming it into a single “name” results in something worse than too many attributes.

OK, I understand the problem, and I don't have a better solution.

Total parameter name count on that geotraces cruise was over 500...

> >I was also wondering about the use of NaN as a _FillValue.  This seems to me a bad practice because it means that code testing for the locations of the _FillValue requires 2 code paths, one to handle the NaN case, and another for the normal case in which a normal value of the correct type is designated.
>
> Use of NaNs for floating point data as recommended by https://www.unidata.ucar.edu/software/netcdf/workshops/2011/bestpractices/MissingData.html


It isn't actually recommended; it is suggested that it be considered. And upon considering it, I think it is a bad recommendation, for the reasons I have stated.  But inevitably it will be followed on occasion.

> xarray appears to coerce non contiguous numeric data into a float type and encode the actual file dtype information for round trip purposes. I don’t know if this is xarray behavior or if the underlying netcdf library forces something like this to happen.

It's xarray; netCDF4 preserves actual data types unless it is doing automatic scaling and offsetting, which is entirely optional (but on by default, if scale and offset attributes are supplied).

> When I was going though the CF mailing list archives for recommendations on fill values, it seemed that the usual practice is to take the approach that xarray does, turn all numeric data into something which follows IEEE 754 rules, then do all the testing/math using nan aware logic. Missing data is a little complicated in netcdf files, for example, I could use the NUG valid_range attribute and not include a _FillValue, then ensure that all “non valid” data remains outside the valid range.

Too many choices; I think it is best to use explicit values (but not NaN) as the _FillValue for missing values.  Then netCDF4-python (and the scipy interface also, though that is just for netcdf3) can automatically generate masked arrays, and C and Fortran code can use normal logic with the recorded _FillValue instead of having to jump through "isnan" hoops for some variables and not for others.  I don't see any value in using valid_range; it is ill-defined and cumbersome.  Ideally, its function might be to provide a hint to plotting routines as to how to scale an axis so as to always include valid values.


> Basically, every approach to missing data values seems to a compromise that no one is happy with.


If everyone could just standardize on the non-NaN _FillValue approach, it would work out fine.  But of course that won't happen.

(Disclosure: if I were still a Matlab user, I would probably prefer the NaN _FillValue approach.  But I would still want to stick with _FillValue, not valid_range or anything else.

Python has much better netcdf interfaces, so the _FillValue choice doesn't matter so much.)


> >What do you think of my suggestion to add a "table of variables" in some form?
>
> I’m having trouble picturing what this would actually look like in the constraints of a netcdf file...
> An array of strings containing values like: “CTDPRS:var0” “CTDSAL:var2”
> Something else maybe?

(Addressed in my previous message.)

Eric

**Comment by DocOtak** [comment]
From @efiring
---
I just looked at the netCDF4-python code and confirmed that, as I expected based on observed behavior, it now includes the necessary special handling for NaN as a _FillValue.  (Or a missing_value.) It also has checks for valid_min, valid_max, and valid_range!

Eric

**Comment by DocOtak** [comment]
I've done a whole lot of looking and cannot seem to find any controlled _variable name_ list anywhere. The closest I can find is the list of [Argo physical parameters](http://www.argodatamgt.org/Documentation). Note that the current document is dated 2018-01-29, I have an inquiry to Thierry regarding any updates to this list, as such, I'm not linking directly to it.

This would result in the following having some standardized _variable name_:

- XYZT Coordinate variables (lon, lat, pressure, date)
- Program Specific Coordinate labels (e.g. expocode, stnnbr, castno, sampno, geotraces_event, geotraces_sample, etc..) actual list is TBD but it *should* be short.
- Some mapping of standard attributes to argo physical parameter names

  - I'd guess that in the event of a physical parameter conflict (e.g. two or more oxygen measurements), the in situ sensor should take precedence to better match argo files. Standard procedure (on GO-SHIP at least) is to calibrate the in situ sensors using bottle samples.
  - The decision to use `<PARAM>` vs `<PARAM>_adjusted` should probably be based on a controlled list of the values the ACDD 1.3 `processing_level` attribute can have. If going off what I've seen in Bob Key's "header" file, the options are "preliminary" or "final". Note that in the Argo data world `_adjusted` means "calibrated" and is what you should be using (in the delayed mode files) if doing any actual science with Argo data.
  - This should probably include the `<PARAM>_qc` for quality flags.

- Everything else would end up with some `var0...varN` name...

**Comment by DocOtak** [comment]
Thierry has confirmed that the Argo physical parameters list dated is 2018-01-29 is the most recent version.

**Comment by DocOtak** [comment]
After a whole bunch of discussions, we will have "standardized" variable names, that db/mapping project is over here: [cchdo/params](https://github.com/cchdo/params)

It seems that almost all the folks we have sent test files to are indeed matlab users and so we'll leave nans for now. We are preparing machinery that will adjust the entire dataset if we do revisit the nan issue, not just for nans, but any format changes we might do.
